sql/expression/arithmetic.go

```diff
@@ -0,0 +1,411 @@
+
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+var (
+	// ErrUnableToCast means that we could not find common type for two arithemtic objects
+	errUnableToCast = errors.NewKind("Unable to cast between types: %T, %T")
+
+	// ErrUnableToEval means that we could not evaluate an expression
+	errUnableToEval = errors.NewKind("Unable to evaluate an expression: %v %s %v")
+)
+
+// Arithmetic expressions (+, -, *, /, ...)
+type Arithmetic struct {
+	BinaryExpression
+	op string
+}
+
+// NewArithmetic creates a new Arithmetic sql.Expression.
+func NewArithmetic(left, right sql.Expression, op string) *Arithmetic {
+	return &Arithmetic{BinaryExpression{Left: left, Right: right}, op}
+}
+
+// NewPlus creates a new Arithmetic + sql.Expression.
+func NewPlus(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.PlusStr)
+}
+
+// NewMinus creates a new Arithmetic - sql.Expression.
+func NewMinus(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.MinusStr)
+}
+
+// NewMult creates a new Arithmetic * sql.Expression.
+func NewMult(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.MultStr)
+}
+
+// NewDiv creates a new Arithmetic / sql.Expression.
+func NewDiv(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.DivStr)
+}
+
+// NewShiftLeft creates a new Arithmetic << sql.Expression.
+func NewShiftLeft(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.ShiftLeftStr)
+}
+
+// NewShiftRight creates a new Arithmetic >> sql.Expression.
+func NewShiftRight(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.ShiftRightStr)
+}
+
+// NewBitAnd creates a new Arithmetic & sql.Expression.
+func NewBitAnd(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.BitAndStr)
+}
+
+// NewBitOr creates a new Arithmetic | sql.Expression.
+func NewBitOr(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.BitOrStr)
+}
+
+// NewBitXor creates a new Arithmetic ^ sql.Expression.
+func NewBitXor(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.BitXorStr)
+}
+
+// NewIntDiv creates a new Arithmetic div sql.Expression.
+func NewIntDiv(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.IntDivStr)
+}
+
+// NewMod creates a new Arithmetic % sql.Expression.
+func NewMod(left, right sql.Expression) *Arithmetic {
+	return NewArithmetic(left, right, sqlparser.ModStr)
+}
+
+func (a *Arithmetic) String() string {
+	return fmt.Sprintf("%s %s %s", a.Left, a.op, a.Right)
+}
+
+// Type returns the greatest type for given operation.
+func (a *Arithmetic) Type() sql.Type {
+	switch a.op {
+	case sqlparser.PlusStr, sqlparser.MinusStr, sqlparser.MultStr, sqlparser.DivStr:
+		if sql.IsInteger(a.Left.Type()) && sql.IsInteger(a.Right.Type()) {
```

Comment:
what if `a.Left` or `a.Right` is `sql.Text`? `'2' + 3` is a valid expression

Reply:
I think we can focus on _number_ right now a extend it by text later. With text you may have different formats like: '1', '1.0', '0x1', '1e-1', etc.

Reply:
Actually it should also work for strings. I've double checked and wrote a test:
```
	require := require.New(t)
	result, err := NewPlus(NewLiteral("2", sql.Text), NewLiteral(3, sql.Float64)).
		Eval(sql.NewEmptyContext(), sql.NewRow())
	require.NoError(err)
	require.Equal(float64(5), result)
```
and it works.

And the reason why is because I use:
```
typ := a.Type()
lval64, err := typ.Convert(lval)
```
what internally uses `ToFloat64E`, ... which has a case:
```
	case string:
		v, err := strconv.ParseFloat(s, 64)
		if err == nil {
			return v, nil
		}
```

Reply:
but because upfront we do not know a type of the expression inside a string we can either rely on the second possible type (left or right, but it also can be a string) or just convert everything what's not predefined to float64
Right now if one of expressions is a text we convert result to float64, even if the second _branch_ can be the `int`, because it can be sth. like `'1.0e-1' + 1`